libafl_qemu/src/hooks.rs

```diff
@@ -459,15 +459,14 @@ where
         invalidate_block: bool,
     ) -> InstructionHookId {
         unsafe {
-            let fat: FatPtr = transmute(hook);
-            GENERIC_HOOKS.push((InstructionHookId(0), fat));
+            let mut fat: Box<FatPtr> = Box::new(transmute(hook));
             let id = self.emulator.set_hook(
-                &mut ((*addr_of_mut!(GENERIC_HOOKS)).last_mut().unwrap().1),
+                transmute(&mut *fat), //transmute satisfy the lifetime
```

Comment:
Wouldn't [into_raw](https://doc.rust-lang.org/std/boxed/struct.Box.html#method.into_raw) be more sound? I _feel_ like keeping a mut borrow around will be UB(?)
In general, `transmute` will cause headaches...

Reply:
Yes mostly, but not in this case, because the `set_hook` function will just convert the mutable ref to a u64, so the mutable reference is not keep. But yes, `into_raw` is probably less likely to break in the future. I'll fix that.

Reply:
`Box::into_raw` is not a possibility here because it consumes the box. I restricted the `transmute` so it don't cause any undesirable side effects, just change the lifetime.

BTW as mentioned on the original issue, you can get a much more sound solution if you pass the index of the hook instead of a ptr, like I did in https://github.com/rbran/LibAFL/commit/b83b8a32fd8c97a4d5a887c3aa4e709f8131dd97. This PR was submitted because it was the easiest to implement. It's ok if you want to refuse it in exchange for an more sound solution.

Reply:
Well the hooks could have a Drop implementation that gets the Box from_raw and then consumes it, or something similar to that. Just throwing ideas around :)

Reply:
> Well the hooks could have a Drop implementation that gets the Box from_raw and then consumes it, or something similar to that. Just throwing ideas around :)

That's actually a good idea, it just requires a significant change in how the code is organized. Currently both the `libafl_qemu` (in rust) and `qemu` (in C) keep track of the hooks, and they never get deleted, making then basically static once they are created. I think this could be possible once we have `emulator_start` and `emulator_stop` implemented into qemu.

Reply:
index based was the implementation till late December, but it has a significant performance loss